### ISSUE TITLE
Enabling a plugin before vernemq starts is impossible

### ISSUE THREAD

**Issue by drf** [erlio/vernemq#216]
Instructions in https://vernemq.com/docs/configuration/plugins.html are pretty clear about enabling plugins, but are missing a very concrete use case: having the plugins being enabled before Verne starts.

This is especially an issue in dockerized environments, where the status must be described and reproduced before the container starts. RabbitMQ allows(-ish) this by issuing

> rabbitmq-plugins enable --offline rabbitmq_management

Which allows building a docker image out of the original one with the desired plugins enabled. vmq-admin, instead, requires the node/cluster to be up before acting on plugins.

Maybe thinking about something like Rabbit's --offline could be a solution?

**Comment by dergraf** [comment]
hi @drf 
you're right this isn't very much streamlined so far and we're hoping to come up with a better solution soon. For the moment you could backup the file `/usr/lib/vernemq/lib/vmq_plugin.conf` which gets updated every time you enable/disable a plugin. If you restart the broker the plugin framework reads this file and sets everything up. In the docker context you could may want to setup a volume pointing to `/var/lib/vernemq` folder, additionally you could maintain the `vmq_plugin.conf` file in this volume and symlink it to `/usr/lib/vernemq/lib`, no?

**Comment by drf** [comment]
Hi @dergraf 
I think that the solution you proposed is the *perfect* one, and needs to be documented. Consider the use case: in docker, we already provide configuration through a docker volume mount (or a kubernetes config map, for example) of vernemq.conf, given we need more verbosity and control than simple environment variables.

In a typical docker scenario, you rarely (maybe never) use vmq-admin, as you expect all of your nodes to come up already provisioned and ready. So mounting vmq_plugin.conf makes complete sense - and if you want, we could maybe add to the docker image a VERNEMQ_ENABLED_PLUGINS env variable so that vmq_plugin.conf gets updated when the docker container starts (not really in favor of that, but might work for many)

tl;dr: Using that file is a great solution, it should be documented. Leaving the issue open just for that.

**Comment by codeadict** [comment]
Agree with @drf , i know this behavior is by design but having the offline parameter that can generate this file without connecting to the server or enabling the plugin itself will be a great improvement, specially working inside containers.

**Comment by kellycampbell** [comment]
One problem with the plugins config being in a directory with all the other libs is it makes it tricky to modify it via a docker volume. Correct me if I'm wrong, but I don't think you can mount just a single file as a "volume" can you? So then you end up mounting the whole lib directory and having to manage copies of the rest of the files and dirs in there.

Why can't the plugin config go in the main vernemq.conf file with everything else?

**Comment by larshesel** [comment]
Hi @kellycampbell. Good questions! Here are, from the top of my head, some of the reasons why the plugin config isn't part of the vernemq.conf:

1) the plugin config is updated by VerneMQ on the fly when enabling or disabling plugins using the CLI tool `vmq-admin` which we currently can't do with the vernemq.conf file
2) it wasn't originally meant for manual manipulation - which is also the reason why it's hidden away in the lib folder like it is, maybe that was a bad decision, but back when that decision was made we didn't think about docker ;) 
3) we consider the format of the file to be internal, not something we want to expose in the config file

So,  either we could move the file somewhere else which might make be an easy fix, but it's unclear, at least to me, where to. The `etc` folder might be an option, but I'm not sure. The other approach would be to incorporate it into the main config file, vernemq.conf, but that would mean we'd have to settle on a stable configuration format and figure out how to make the vernemq.conf updateable on the fly (and preserve manual changes). It's unclear how much effort we'd have to put into a change like that, so as long as there's a reasonable alternative we'd rather avoid this.

I don't know if you can mount a single file as a volume (I know little about docker), but wouldn't the symlink trick mentioned above work for you?

**Comment by codeadict** [comment]
@larshesel i guess RabbitMQ updates the `conf` file on **/etc/rabbitmq/rabbitmq.conf**, when one enables a new plugin.

**Comment by kellycampbell** [comment]
From my research, I couldn't find a clear way to do symlinks in docker. Maybe I'm just missing something. I'm using Kubernetes, and I worked around this and the env var with dots problem with a ConfigMap and some extra bits in the start script. https://github.com/kellycampbell/docker-vernemq/tree/k8s-fixes

I didn't want to put the whole vernemq.conf file in the config map. I just wanted to have the overrides, so that's why the extra mount point /etc/vernemq-cfg. This seems to work pretty well, and limits the special cases for my setup to just the start script, so the docker image is still usable outside of k8s.

I'm basically trying to do almost the exact same thing as https://github.com/erlio/vernemq/issues/262

**Comment by kellycampbell** [comment]
Looks like I was thinking too much on the docker symlinks thing. You can just do 

`RUN ln -s ...`

In addition to VerneMQ native plugins, we'd probably also want to be able to configure lua script and webhook plugins via a config file before startup.

At the bottom or the Warning on https://vernemq.com/docs/administration/ it even says this:

> vmq-admin gives you the flexibility to test stuff and react live, but please persistent any static configuration you need in the vernemq.conf file.

**Comment by larshesel** [comment]
We're a bit inconsistent in regard to that warning as some things do get persisted. We've started some work to clean this up for plugins (so you can enable them directly in the vernemq.conf file) making them consistent with this warning, see #265. It's a work in progress, of course.

Currently the `vmq_diversity` (lua) and `vmq_webhooks` persist their data into the `priv` folder of the plugin. For `vmq_webhooks` the file is called `vmq_webhooks.conf` and for `vmq_diversity` the lua scripts can be dumped into the `scripts` folder and they'll be loaded when the plugin starts.

We might, hovewer, make all this configurable from the `vernemq.conf` file too.

**Comment by kellycampbell** [comment]
I think just having those extra files documented and being able to override their location from the vernemq.conf file is probably easier and sufficient.

**Comment by dergraf** [comment]
fixed in #265
